src/box/relay.cc

```diff
@@ -472,6 +475,16 @@ relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,
 		relay_delete(relay);
 	});
 
+	if (unlikely(cursor_engine != NONE_ENGINE)) {
+		relay_initial_file_join(relay->io, &relay->stream, sync,
+					vclock, replica_version_id,
+					cursor_engine, cursor_vclock,
+					cursor_lsn);
```

Comment:
I don't like that there's the new "file join" code path. Basically, it's normal join with the resume position. Can you pass these arguments to the engine's join callback and let the engine handler resume join instead of restarting it?

Reply:
I'll discuss this with Sergey Petrenko. Please, don't resolve the comment

Reply:
> Basically, it's normal join with the resume position.

But it doesn't have same logic: instead of syncing wal we find needed vclock from gc checkpoints, we don't waiting for limbo confirmation, we don't need to prepare read-views. The only thing in common - sending meta stage. Moreover, we need the shared context in case of file join in order to continue from the same LSN in vinyl.

I can add cursor to `engine_join_ctx` structure and pass it `engine_join_ctx`, but it doesn't seem it'll be better, due to a huge number of IFs in `relay_initial_join`, `engine_prepare/complete_join`, `engine_join_xc`.

Let's see, what others will say. IMHO, the separate path for file join and separate `checkpoint_initial_join` is better. @Gerold103 @sergepetrenko

Reply:
I agree with Vova, looks like you could even reduce some code duplication if you make it at the engine level. Huge number of `if`s - how many? Lets give this way a chance, maybe it won't look bad?

Reply:
Done this. scoped guards looks way to ugly